### PR TITLE
support for softnpu-to-softnpu links

### PR DESCRIPTION
**Pull request by rcgoodfellow**
## Primary Change

This patch adds a `Runner::softnpu_links` method which allows softnpu ports to be directly connected.

## Other updates/fixes

* `node!` macro

Add a `node!` macro to reduce repetitiveness when specifying nodes
```rust
let foo = d.node("foo", "helios-1.2", 4, gb(4));
```
becomes
```rust
node!(foo, "helios-1.2", 4, gb(4))
````
which expands to the former.

* remove `println!` usage in favor of logging

* fix `NAME_REGEX`

This regex was matching any string that _contains_ a valid name. So an invalid name containing a valid name would pass. Added `^`/`$` to the beginning/ending of the regex to ensure that an entire string must be a valid name.

### CODE REVIEW

File: lib/src/lib.rs [hunk 1]
```diff
@@ -27,6 +27,13 @@ use std::process::Command;
 use std::str::FromStr;
 use tokio::time::{sleep, Duration};
 
+#[macro_export]
+macro_rules! node {
+    ($d:ident, $name:ident, $img:literal, $cores:literal, $mem:expr) => {
+        let $name = $d.node(stringify!($name), $img, $cores, $mem);
+    };
+}
+
```
Comment (internet-diglett) [inline lib/src/lib.rs:36]: Just need a little clarification. In the example statement you provided, you show:
```
node!(foo, "helios-1.2", 4, gb(4))
```
Where does the macro get the identifier `$d` from?
I'm guessing this is just a typo and the example should be
```
node!(d, foo, "helios-1.2", 4, gb(4))
```

File: lib/src/lib.rs [hunk 2]
```diff
@@ -320,6 +342,36 @@ impl Runner {
...
+    ) -> LinkRef {
+        let r = LinkRef {
+            _index: self.deployment.links.len(),
+        };
+        let l = Link {
+            endpoints: [
+                Endpoint {
+                    node: node1,
+                    index: self.deployment.nodes[node1.index].radix,
+                    kind: EndpointKind::SoftNPU(mac1),
```
Comment (internet-diglett) [inline lib/src/lib.rs:360]: In the event that `mac1` or `mac2` is `None`, what happens?

File: lib/src/lib.rs (hunk 1 above)
Comment (rcgoodfellow) [reply lib/src/lib.rs:36]: yup good catch, a typo indeed. Will correct.

Comment (rcgoodfellow) [reply lib/src/lib.rs:36]: I'm useless without a compiler haha

Comment (internet-diglett) [reply lib/src/lib.rs:36]: Lol, that's why we like Rust! 🤣

File: lib/src/lib.rs (hunk 2 above)
Comment (rcgoodfellow) [reply lib/src/lib.rs:360]: This means that no mac will be passed to the propolis config.

Back when the propolis softnpu device would create a viona guest interface per softnpu port, if a mac was explicitly set that mac would show up in the guest, and the propolis softnpu machinery would generate a random one.

However, today the propolis softnpu device no longer creates a viona guest device per port. We rely on the tfport kernel driver to do that, and only supply a single port that carries sidecar-header encapsulated ports to the guest.

So tl;dr; - with the current propolis softnpu implementation, these mac options do nothing. I'll file a bug to fix this.

Comment (internet-diglett) [reply lib/src/lib.rs:360]: Thanks for the in depth explanation! That clears things up a lot!
